### ISSUE TITLE
Does distiller need the PyTorch CNN to be defined in some fixed ways?

### ISSUE THREAD

**Issue by hilman-dayo** [NervanaSystems/distiller#318]
Hello.
First of all, thank you for this awesome library.

I have one question.
This question might sound simple and "noob". However, I really need to ask this still I am still quite a beginner in this domain (especially in trying to shrink the CNN size) and need some guidance.

I will build a custom CNN architecture with PyTorch. Because I have only worked with Keras in the past, I am still vague with the PyTorch framework in defining a CNN model. That being said, I am learning PyTorch and am trying to build my own custom CNN right now.
The thing is, I need to make sure that my custom CNN can be shrunk/compress with the distiller library. Is it safe for me to assume that distiller can work with any arbitrary CNN model defined in PyTorch? Or does distiller needs the CNN model to be defined in some fixed way?

Actually, I already checked and executed the compress_classifier.py file but I still can't really tell on how the distiller actually interacts with the CNN model and this leads me on asking this question.

Thank you.

**Comment by levzlotnik** [comment]
Hi @hilman-dayo,

That mostly depends on how you want to compress your model.  
If you want to prune weights, as long as the weights are seen by `model.parameters()` - you're good to go. In PyTorch, the model weights (and in general any trainable tensors) are best defined as `nn.Parameter` objects, which is a subclass of `torch.Tensor` which knows that inside a module - it usually serves as a weight tensor. More info in [pytorch `nn.Parameter` docs](https://pytorch.org/docs/stable/nn.html#torch.nn.Parameter).

If you want to quantize the model, you should add every operation in the `.forward` function as a layer in the model, and then call this layer!
E.g. if the regular forward is done with residuals like
```
def forward(self, x):
    x_res = x
    x = self.conv1(x)
    x = self.bn1(x)
    x = x + x_res
    return x
```
Instead - in the module constructor you should add a layer for element wise addition, so this element wise addition may be quantized by our quantizers, like this
```
def __init__(self, ...):
    ... # previous/regular init
    self.eltwiseadd_residual = distiller.modules.EltwiseAdd()

def forward(self, x):
    x_res = x
    x = self.conv1(x)
    x = self.bn1(x)
    x = self.eltwiseadd_residual(x, x_res)
    return x
```
This way, you can assure the quantizer will find all of the modules inside the model graph and be able to quantize properly every operation :)  
Let us know if you have more questions.

Lev

**Comment by hilman-dayo** [comment]
Thank you for the answer.
It really helped.

One other question.
As I said previously, I want to build my own custom Object Detection framework and make it smaller with distiller.
In the future, I want to convert it into an ONNX model.
Will the fact that my custom Object Detection framework being pruned etc. with distiller make it is impossible for it to be exported to ONNX model?

Is there anything that I need to keep in mind whenever I am designing and pruning my model so that it is possible to be converted to ONNX model in the future?

**Comment by levzlotnik** [comment]
Hi @hilman-dayo 

No, you don't have to worry about exporting to onnx compressed models.
You can export pruned models already right now - since the weights are the only thing that are changed in the model, and not the structure itself. Not quite sure about thinned models, I assume it would work as well.

Quantized models are a bit harder - we simulate the quantization, and when exporting to onnx you would actually see the simulation instead of actual quantization ops. We are thinking of implementing a way to export a quantized model, but we're still not quite there in terms of work schedule.

In short - no, you don't have to worry about export to onnx, eventually we'll support everything.

Thanks,
Lev

**Comment by hilman-dayo** [comment]
Thank you for the answer. I may close this thread right now.
